### ISSUE TITLE
Postprocessor function fires for every matched character

### ISSUE THREAD

**Issue by matthew-dean** [kach/nearley#383]
One of the things that's unexpected for me as I'm testing is that, if the post processor is matching a sequence, it fires for every matched identifier, including partial matches. Is that expected? For example, if I'm matching the `:extend`, and I do `{% d => console.log(d) %}`, I get logs like (see the second item in the nested array): 
![screen shot 2018-06-25 at 6 36 12 pm](https://user-images.githubusercontent.com/414752/41884223-55fc6296-78a7-11e8-90e0-5d19854469a0.png)

It's not an ambiguous grammar problem (I presume), because it produces one "parsing". I was naively creating objects on each post-processor call, but is this behavior expected? Should I be looking for a "Ok are you really done?" flag? Is this for error handling? The brief documentation doesn't give any indication whatsoever that partial matches of a Nearley rule will fire the post-processor, and I don't want to create objects that are never used. Can you explain?

**Comment by matthew-dean** [comment]
From the docs:

> For this purpose, each rule can have a postprocessor: a JavaScript function that transforms the array and returns a “processed” version of the result.

My presumption wouldn't be that a new array is created per character and calls the postprocessor function every time, since that would be terribly inefficient. I presume this a problem with my grammar? Or did I trigger a bug?

**Comment by kach** [comment]
No, no bug, this is how it's supposed to work. Remember that nearley is a _streaming_ parser — it never knows when you're done feeding input. So if you feed it `:extend`, and it has only seen `:exte` so far, it sees that as a valid "pseudo-selector" and runs the postprocessor on it. In general, we _strongly_ recommend not having side-effects in postprocessors.

**Comment by kach** [comment]
We considered setting it up so that all postprocessors run at the end, but for a couple of reasons that would be a backwards-incompatible change. For now, if you worry about efficiency at that level, really the first thing to do is to use a moo lexer.

**Comment by jtenner** [comment]
I was trying to build a parser that had side-effects and was curious to venture down this road.  Specifically, it was for AST node caching. The language was designed for a binary calculation tree, and I was just afraid to dip my toes in the water. Thanks for the encouragement!

**Comment by matthew-dean** [comment]
@kach 

> Yeah, a lexer would prevent this because :extend would be read by nearley in one chunk.

Ohhhhh really?? That's reason enough for me to invest the time in researching `moo` / lexers.

**Comment by jtenner** [comment]
Thanks for the information. I'm using the words that I know, and happy to learn.
